src/internal.c

```diff
@@ -22357,12 +22364,50 @@ static int CheckTLS13AEADSendLimit(WOLFSSL* ssl)
     }
 
     if (w64GTE(seq, limit))
-        return Tls13UpdateKeys(ssl); /* Need to generate new keys */
+        return Tls13UpdateKeys(ssl, 1); /* Need to generate new keys */
 
     return 0;
 }
 #endif /* WOLFSSL_TLS13 && !WOLFSSL_TLS13_IGNORE_AEAD_LIMITS */
 
+#if defined(WOLFSSL_TLS13) && defined(WOLFSSL_TLS13_AUTO_REKEY)
+/**
+ * Set rekey required if sequence number is equal to 2^maxInv.
+ *
+ * ssl     SSL/TLS object.
+ * maxInv  Maximum number of invocations of cipher. Count = 2^maxInv.
+ * hi      Next hi 32-bits of sequence number.
+ * lo      Next lo 32-bits of sequence number.
+ * sub     Amount to subtract from sequence number.
+ * returns 0 when no rekey required and 1 when rekey required.
+ */
+static int Tls13_NeedRekey(WOLFSSL* ssl, byte maxInv, word32 hi, word32 lo,
+    word32 sub)
+{
+    if (ssl->options.rekey != REKEY_NONE)
+        return 0;
+
+    hi -= (lo < sub) ? 1 : 0;
+    lo -= sub;
+
+    if (maxInv == 0 || maxInv == 64) {
+        if (hi == 0xffffffff && lo == 0xffffffff) {
+            return 1;
+        }
+    }
+    else if (maxInv < 32) {
+        if (lo == ((1U << maxInv) - 1)) {
+            return 1;
+        }
+    }
+    else if ((hi == ((1U << (maxInv - 32)) - 1)) && (lo == 0xffffffff)) {
+        return 1;
+    }
```

Comment:
Please document this section better. I'm having a hard time wrapping my head around what the intention here is. Consider using the `w64wrapper` to make this logic simpler. Also It would be better to use `>=` than `==` to catch any edge cases where we miss a check.